src/lib.rs

```diff
@@ -1,101 +1,405 @@
-                ),
-            );
+    ])
+}
+
+fn write_files<'a>(files_to_write: impl Iterator<Item = (&'a Path, &'a [u8])>) {
+    for (path, file_contents) in files_to_write {
+        fs::write(path, file_contents).unwrap();
+    }
+}
+
+type RuleName = String;
+
+fn apply_fixes(file_contents: &mut Vec<u8>, pending_fixes: HashMap<RuleName, Vec<PendingFix>>) {
+    let non_conflicting_sorted_pending_fixes =
+        get_sorted_non_conflicting_pending_fixes(pending_fixes);
+    for PendingFix { range, replacement } in non_conflicting_sorted_pending_fixes.into_iter().rev()
+    {
+        file_contents.splice(range, replacement.into_bytes());
+    }
+}
+
+fn compare_pending_fixes(a: &PendingFix, b: &PendingFix) -> Ordering {
+    if a.range.start < b.range.start {
+        return Ordering::Less;
+    }
+    if a.range.start > b.range.start {
+        return Ordering::Greater;
+    }
+    if a.range.end < b.range.end {
+        return Ordering::Less;
+    }
+    if a.range.end > b.range.end {
+        return Ordering::Greater;
+    }
+    Ordering::Equal
+}
+
+fn has_overlapping_ranges(sorted_pending_fixes: &[PendingFix]) -> bool {
+    let mut prev_end = None;
+    for pending_fix in sorted_pending_fixes {
+        if let Some(prev_end) = prev_end {
+            if pending_fix.range.start < prev_end {
+                return true;
+            }
+        }
+        prev_end = Some(pending_fix.range.end);
+    }
+    false
+}
+
+fn get_sorted_non_conflicting_pending_fixes(
+    pending_fixes: HashMap<RuleName, Vec<PendingFix>>,
+) -> Vec<PendingFix> {
+    pending_fixes.into_iter().fold(
+        Default::default(),
+        |accumulated_fixes, (rule_name, mut pending_fixes_for_rule)| {
+            pending_fixes_for_rule.sort_by(compare_pending_fixes);
+            if has_overlapping_ranges(&pending_fixes_for_rule) {
+                panic!("Rule {:?} tried to apply self-conflicting fixes", rule_name);
+            }
+            let mut tentative = accumulated_fixes.clone();
+            tentative.extend(pending_fixes_for_rule);
+            if has_overlapping_ranges(&tentative) {
+                accumulated_fixes
+            } else {
+                tentative
+            }
         },
     )
-    .unwrap();
-    if reported_any_violations.load(Ordering::Relaxed) {
-        process::exit(1);
-    } else {
-        process::exit(0);
+}
+
+#[derive(Default)]
+struct AllPendingFixes(Mutex<HashMap<PathBuf, PerFilePendingFixes>>);
+
+impl AllPendingFixes {
+    pub fn append(
+        &self,
+        path: &Path,
+        file_contents: &[u8],
+        rule_name: &str,
+        fixes: Vec<PendingFix>,
+    ) {
+        self.lock()
+            .unwrap()
+            .entry(path.to_owned())
+            .or_insert_with(|| PerFilePendingFixes::new(file_contents.to_owned()))
+            .pending_fixes
+            .entry(rule_name.to_owned())
+            .or_default()
+            .extend(fixes);
+    }
+
+    pub fn into_inner(
```

Comment:
Without this when I tried to call `.into_inner()` on an `AllPendingFixes` it understandably complained that I couldn't take ownership of the wrapped type via a `Deref` implementation

So this worked but I don't know if there's a better way (eg would `DerefMut` suffice)?